docs/ha-considerations.md

```diff
@@ -0,0 +1,337 @@
+# High Availability Considerations
+
+## Overview
+
+When setting up a productive cluster, high availability, i.e. the cluster's ability to remain operational even if some control plane or worker nodes fail, is usually a requirement. For worker nodes, assuming that there are enough of them, this is part of the very cluster functionality; however redundancy of control plane nodes and `etcd` instances needs to be catered for when planning and setting up a cluster.
+
+`kubeadm` supports setting up of multi control plane and multi `etcd` clusters (see [Creating Highly Available clusters with kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/), part of the official Kubernetes documentation, for a step-by-step guide). However there are some more things to consider and set up which are not part of Kubernetes itself and hence not covered there. This document provides some additional information and examples useful when planning and bootstrapping HA clusters with `kubeadm`.
+
+<!-- TOC -->
+
+- [High Availability Considerations](#high-availability-considerations)
+    - [Overview](#overview)
+    - [Options for Software Load Balancing](#options-for-software-load-balancing)
+        - [keepalived and haproxy example](#keepalived-and-haproxy-example)
+        - [kube-vip example](#kube-vip-example)
+        - [Bootstrap the cluster](#bootstrap-the-cluster)
+
+## Options for Software Load Balancing
+
+When setting up a cluster with more than one control plane, one will typically put the API Server instances behind a load balancer and use the `--control-plane-endpoint` option when running `kubeadm init` for the new cluster to use it.
+
+Of course, the load balancer itself should be highly available, too. This is usually achieved by setting up more load balancers in a cluster of hosts managing a virtual IP, so that always the load balancer on the host currently holding the vIP will be used.
```

Comment:
> This is usually

do we know for sure that this is the more common way out there?

> setting up more load balancers in a cluster of hosts managing a virtual IP

i don't think that if we deploy multiple keepalived instances on the nodes we should say that there are multiple loadbalancer for that cluster. it's still a single load balancer with multiple instances.

Reply:
> > This is usually
> 
> do we know for sure that this is the more common way out there?

I'm rewording this a bit to "higher availability can be achieved", that's more neutral. 

> > setting up more load balancers in a cluster of hosts managing a virtual IP
> 
> i don't think that if we deploy multiple keepalived instances on the nodes we should say that there are multiple loadbalancer for that cluster. it's still a single load balancer with multiple instances.

Rewording this, too. Let's see if it's better.